src/internal/orm/orm.go

```diff
@@ -0,0 +1,68 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package orm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/astaxie/beego/orm"
+	"github.com/goharbor/harbor/src/common/utils/log"
+)
+
+type ormKey struct{}
+
+// FromContext returns orm from context
+func FromContext(ctx context.Context) (orm.Ormer, bool) {
+	o, ok := ctx.Value(ormKey{}).(orm.Ormer)
+	return o, ok
+}
+
+// NewContext returns new context with orm
+func NewContext(ctx context.Context, o orm.Ormer) context.Context {
+	return context.WithValue(ctx, ormKey{}, o)
+}
+
+// WithTransaction a decorator which make f run in transaction
+func WithTransaction(f func(ctx context.Context) error) func(ctx context.Context) error {
+	return func(ctx context.Context) error {
+		o, ok := FromContext(ctx)
+		if !ok {
+			return fmt.Errorf("ormer value not found in context")
+		}
+
+		tx := ormerTx{Ormer: o}
+		if err := tx.Begin(); err != nil {
+			log.Errorf("begin transaction failed: %v", err)
+			return err
+		}
+
+		if err := f(ctx); err != nil {
+			if e := tx.Rollback(); e != nil {
```

Comment:
Suggest wrapping the error and then return it. Overriding error or discard error may lose some error context.

Reply:
Leverage our common error

Reply:
Not everything needed to be common error, this is only the raw error from the sql driver, i think it's ok for now.

The API error handler will transform this error to 500 internal server error.

Reply:
IMO we only need to make sure the error is properly wrapped before the API returns, otherwise when the code stack is deep it will be wrapped more than necessarily.

Reply:
```go
func load(){
if err:=call(); err!=nil{
  return errors.Wrap(err, "my context for load") // then the error string will be `error string : my context for load`; easy for tracing error becasue error may happen in diffeent caller contexts
}
}

This is just a suggestion, not a required term
depends on you

```